File: go/01_les_bases/bases.go
/*
 * MODULE 01 - Les briques de base en Go
 * =====================================
 * Illustre, dans l'ordre, les notions du README : variables et types,
 * fmt.Println / fmt.Printf, conditions, boucle for (ses 3 formes), fonctions.
 *
 * ⚠️ Ce fichier a SON PROPRE func main : on le lance SEUL.
 * Lancer :
 *     go run go/01_les_bases/bases.go
 */

package main

import "fmt" // boîte à outils d'affichage (Println, Printf)

// ─────────────────────────────────────────────
// FONCTIONS
// (en Go, elles peuvent être définies au-dessus OU en dessous de main)
// ─────────────────────────────────────────────

// (a int, b int) = deux paramètres entiers ; le 'int' APRÈS = type de retour.
func additionner(a int, b int) int {
	return a + b // 'return' renvoie le résultat à l'appelant
}

// Pas de type après les parenthèses = cette fonction ne renvoie rien.
func saluer() {
	fmt.Println("Bonjour depuis une fonction !")
}

// ─────────────────────────────────────────────
// main : le point de départ du programme
// ─────────────────────────────────────────────
func main() {

	// 1. VARIABLES ET TYPES
	var age int = 30 // forme LONGUE : var, nom, type (int), valeur
	taille := 1.68   // forme COURTE : Go devine le type (float64)
	nom := "Sam"     // une chaîne de caractères (string), entre guillemets "
	majeur := true   // un booléen (bool) : true ou false

	// 2. AFFICHER
	//    Println = simple, ajoute un saut de ligne, sépare par des espaces.
	fmt.Println("Nom :", nom)
	//    Printf = contrôle précis ; chaque %... = une valeur ; \n = saut de ligne.
	fmt.Printf("Age : %d ans\n", age)       // %d = entier
	fmt.Printf("Taille : %.2f m\n", taille) // %f = décimal ; .2 = 2 décimales
	fmt.Printf("Majeur : %t\n", majeur)     // %t = booléen

	// 3. CONDITIONS : PAS de parenthèses, mais accolades { } obligatoires.
	note := 12
	if note >= 16 {
		fmt.Println("Mention : Tres bien")
	} else if note >= 10 { // 'else if' (et non 'elif')
		fmt.Println("Mention : Recu")
	} else {
		fmt.Println("Mention : A retravailler")
	}

	// 4. BOUCLE for, forme 1 : (début ; condition ; pas) — comme le for du C.
	for i := 0; i < 3; i++ { // i++ = i = i + 1
		fmt.Println("Tour numero", i)
	}

	// 5. BOUCLE for, forme 2 : une seule condition = le "while" des autres langages.
	compteur := 0
	for compteur < 3 {
		fmt.Println("compteur =", compteur)
		compteur++ // indispensable, sinon boucle infinie
	}

	// 6. BOUCLE for, forme 3 : sans condition = infinie, arrêtée par "break".
	n := 0
	for {
		if n == 2 {
			break // stoppe la boucle
		}
		fmt.Println("n =", n)
		n++
	}

	// 7. APPELER NOS FONCTIONS.
	saluer()
	somme := additionner(7, 5)
	fmt.Printf("7 + 5 = %d\n", somme)
}

File: rust/01_les_bases/bases.rs
/*
 * MODULE 01 - Les briques de base en Rust
 * =======================================
 * Illustre, dans l'ordre, les notions du README : variables (let / let mut),
 * types, if/else (et if comme expression), boucles (loop/while/for), fonctions.
 *
 * Compiler puis lancer :
 *     rustc rust/01_les_bases/bases.rs -o bases && ./bases
 */

// ─────────────────────────────────────────────
// FONCTIONS (on peut les définir avant OU après main, Rust s'y retrouve)
// ─────────────────────────────────────────────

// 'fn' = fonction. (a: i32, b: i32) = deux paramètres entiers.
// '-> i32' = le type de la valeur RENVOYÉE.
fn additionner(a: i32, b: i32) -> i32 {
    a + b   // dernière ligne SANS ';' = c'est la valeur renvoyée (pas besoin de 'return')
}

// Sans '->', la fonction ne renvoie rien : elle se contente d'afficher.
fn saluer() {
    println!("Bonjour depuis une fonction !");
}

// ─────────────────────────────────────────────
// main : le point de départ du programme
// ─────────────────────────────────────────────
fn main() {

    // 1. VARIABLES ET TYPES
    //    'let' crée une variable. Le type est souvent DEVINÉ, mais on peut l'écrire après ':'.
    let age: i32 = 30;        // entier
    let taille: f64 = 1.68;   // nombre à virgule
    let initiale: char = 'A'; // UN caractère, entre apostrophes simples
    let actif: bool = true;   // vrai / faux
    let prenom: &str = "Alice"; // un texte fixe

    // 2. AFFICHER avec println! : chaque {} est remplacé par une variable, dans l'ordre.
    println!("Prenom : {}", prenom);
    println!("Age : {} ans", age);
    println!("Taille : {} m", taille);
    println!("Initiale : {}", initiale);
    println!("Actif : {}", actif);

    // 3. IMMUABLE PAR DÉFAUT, sauf si on ajoute 'mut'.
    //    Sans 'mut', on ne pourrait PAS faire "compteur = ..." plus loin.
    let mut compteur = 0;   // 'mut' = on a le droit de la modifier

    // 4. CONDITIONS : la condition SANS parenthèses, le bloc entre { }.
    let note = 12;
    if note >= 16 {
        println!("Mention : Tres bien");
    } else if note >= 10 {   // 'else if' (et non 'elif')
        println!("Mention : Recu");
    } else {
        println!("Mention : A retravailler");
    }

    // 5. if EST UNE EXPRESSION : il peut renvoyer une valeur rangée dans un 'let'.
    let reussi = note >= 10;
    let message = if reussi { "gagne" } else { "perdu" };
    println!("Resultat : {}", message);

    // 6. BOUCLE for : pour chaque i dans la plage 0..3 (0, 1, 2 ; le 3 est EXCLU).
    for i in 0..3 {
        println!("Tour numero {}", i);
    }

    // 7. BOUCLE while : tant que la condition est vraie.
    while compteur < 3 {
        println!("compteur = {}", compteur);
        compteur += 1;   // indispensable, sinon boucle infinie
    }

    // 8. BOUCLE loop : infinie par nature ; on en sort avec 'break'.
    let mut tours = 0;
    loop {
        if tours == 2 {
            break;   // on quitte la boucle
        }
        println!("loop, tour {}", tours);
        tours += 1;
    }

    // 9. APPELER NOS FONCTIONS définies plus haut.
    saluer();
    let somme = additionner(7, 5);
    println!("7 + 5 = {}", somme);
}
